src/com/selftest/Stack.java

```
@@ -0,0 +1,91 @@
+package com.selftest;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class Stack implements Iterable<Character> {
+    private char[] stack;
+    private int capacity;
+    private int size;
+
+    public Stack(int capacity) {
+        this.capacity = capacity;
+        this.stack = new char[this.capacity];
+        this.size = 0;
+    }
+
+    public Stack(Stack other) {
+        this.capacity = other.capacity;
+        this.stack = new char[this.capacity];
+        IntStream.range(0, other.size).forEach(i -> this.push(other.stack[i]));
+    }
+
+    public Stack(char[] values) {
+        this.capacity = values.length;
+        this.stack = new char[this.capacity];
+        for (char c : values)
+            this.push(c);
+    }
+
+    @Override
+    public Iterator<Character> iterator() {
+        return new Iterator<Character>() {
+            @Override
+            public boolean hasNext() {
+                return !isEmpty();
+            }
+
+            @Override
+            public Character next() {
+                return pop();
+            }
+        };
+    }
+
+    public Stream<Character> stream() {
+        return StreamSupport.stream(this.spliterator(), false);
+    }
+
+    public Stack push(char value) throws IndexOutOfBoundsException {
+        if (size + 1 > capacity)
+            throw new IndexOutOfBoundsException("Error: The stack is full.");
+
+        this.stack[size++] = value;
+        return this;
+    }
+
+    public char pop() throws NoSuchElementException {
+        if (isEmpty())
+            throw new NoSuchElementException("Error: The stack is empty.");

```

This exception type is typically used for a search failure rather than an empty container situation. Some kind of underflow exception type would be a good choice.